### ISSUE TITLE
Add functionality to check if all images have alt text

### ISSUE THREAD

**Issue by spcanelon** [thisisnic/docreview#38]
👋🏽 Hi @thisisnic! I saw [your recent tweet](https://twitter.com/nic_crane/status/1427543856967528453?s=20) and was excited to learn about your package, it looks so useful! Thank you for being open to feature requests 🙂 

## Feature

I think it would be helpful if `docreview` could help developers check (by default) that all images contain a text description, also known as alternative (alt) text. [Alt text can be added to images in multiple ways using R Markdown](https://blog.rstudio.com/2021/04/20/knitr-fig-alt/) and ends up hard-coded in the HTML behind the scenes as something like: 

```html
<img src="image.png" alt="image alt text">
```

## Motivation

Effective alt text ensures that the image is made accessible to folks that use screen readers. These are [usually folks with visual disabilities (e.g. blind, low vision), but not exclusively](https://webaim.org/projects/screenreadersurvey9/). Alt text also appears in place of an image when the image won't load on a website because of internet limitations or something else.

Package documentation pages often contain images that communicate something important about the package. When these images don't have alt text, it makes the information they are meant to convey inaccessible to R users that use screen readers. Images without alt text are often read aloud by default as "image" which you can imagine isn't helpful when trying to learn how to use a package.

## Resources

I have a curated list of resources that I can share, but I didn't want to prematurely overwhelm the issue so here are just a few 😅 

- [WebAIM: Alternative Text](https://webaim.org/techniques/alttext/)
- [WebAIM: Screen Reader User Survey #9 Results](https://webaim.org/projects/screenreadersurvey9/)
- [Accessible Data Science Beyond Visual Models: Non-Visual Interactions with R and RStudio Packages](https://www.rstudio.com/resources/rstudioglobal-2021/accessible-data-science-beyond-visual-models-non-visual-interactions-with-r-and-rstudio-packages/) - RStudio Global 2021 talk by JooYoung Seo

---

Thanks for reading! I'm more than happy to keep the conversation going and/or share more specific resources (e.g. alt text for data viz, how to write alt text, etc)

**Comment by thisisnic** [comment]
I think this is a fantastic idea, and thank you for putting together such an in-depth feature request! Are you interesting in making a PR or just wanting to request the feature? Either is good!

I'm wondering if it would be sufficient to check that any images contain alt text, or, if there's any other best practices we could incorporate (e.g. min sentence length for alt text?)

I've seen packages which automatically generate alt text for data viz outputs, which is beyond the scope of this package, but if there's anything specific that you think could be added, let me know. On the other hand, there may be too much chance of false positives with making it overcomplicated, and just checking for alt text existing at all would be sufficient to help.

This might be a significant change as - as you said - there are multiple ways of including images so to make sure we caught all images, we'd perhaps have to implement something to compile the docs and then scan the output.

**Comment by spcanelon** [comment]
Yay! I'm excited to hear you think it's a good idea! My intention was to just request the feature -- I don't know enough about package building to be able to make a PR 😅 Your approach to maybe compile the docs and then scan the output makes sense to me! So checking the contents of all `alt` attributes in `img` HTML tags?

I agree with you that just checking for alt text would be ideal! Also, automatically-generated alt text for data visualizations is a complicated topic and an area that I wouldn't wade into without inclusive and fair participation from folks with visual disabilities. 

I've learned that unfortunately it is all too common for non-disabled folks to come up with technological "fixes" that don't center the people most affected (disregarding the saying "nothing about us, without us") and that are inadequate as a result 😞
Multiple branches of [this twitter thread](https://twitter.com/AmyCesal/status/1395002388734427136?s=20) capture some of the nuance. I'll link to a few specific comments/replies from Liz Hare who provides perspective as an experienced and blind R user and who has done a lot of accessibility work within the R community:

- https://twitter.com/DogGeneticsLLC/status/1395049458589655045?s=20
- https://twitter.com/DogGeneticsLLC/status/1395029209647566850?s=20
- https://twitter.com/DogGeneticsLLC/status/1395381149942390786?s=20

**Comment by thisisnic** [comment]
Thanks for the extra context there @spcanelon! :smiley:   The development version of `docreview` now contains this functionality.  I've tested it a little, but if you'd find it useful and would like to test it on some more packages, that'd be a great help.

(Also, your use of emojis has inspired me to add some to my README on the current branch I'm working on, so thanks for that too!)

**Comment by spcanelon** [comment]
You are so very welcome, thank you for appreciating it! 😃

And wow, that was such a fast turnaround, thank you so much! 🎉 

I would be more than happy to test the functionality myself but it might take some time because [the one package](https://github.com/spcanelon/TidyTuesdayAltText) I've developed does not have its own documentation site (yet!). Would you be open to me sharing on Twitter in an effort to find some testers?

(Your comment about the emojis made me smile! I think they're fun and I've enjoyed learning about [how to use them with accessibility in mind](https://uxdesign.cc/emojis-in-accessibility-how-to-use-them-properly-66b73986b803))

**Comment by sharlagelfand** [comment]
Coming over here from twitter spying 🔍 I would be happy to test on a couple of my packages! Especially since I know they don't have alt text (yikes) so will get to see how it looks before and after!

**Comment by thisisnic** [comment]
@spcanelon That would be great!  I've now set up a CI job which runs the code against the most recently updated r-lib packages, but unless I trawl through everything manually, I won't really know if there are any false negatives, so that would be amazing! :pray: 

@sharlagelfand Fabulous, thanks mate! :heart: 

I'll pop some instructions on how to test this below, so we can direct any testers to them.

**Comment by thisisnic** [comment]
## Instructions

Thanks for helping with testing the new alt-text checking feature of `docreview`!  Below are instructions for how to test this feature.

1. Render the package vignettes to HTML
    - you can do this via `tools::buildVignettes(dir=<path to your package root directory>)` or by opening each vignette in RStudio and clicking the "Knit" button.

2. Install the dev version of `docreview` (which contains the alt text checking functionality)
    - run `remotes::install_github("thisisnic/docreview")` in R

3. Run `docreview` against your package and check the results printed in the console
    - run `docreview::package_review(<path to your package root directory>)` in R

If your package contains vignettes that have been rendered as HTML (as per instructions in step 1), the output should contain a message about how many images there are in your vignettes, and how many of these images contain alt text.  Please check that the values reported are accurate!

If you find any issues or problems, please [file an issue](https://github.com/thisisnic/docreview/issues) containing either a [reproducible example](https://github.com/tidyverse/reprex) or a link to the git repo for the package you're testing this against (a commit number would be super helpful too!)

Thank you!

**Comment by simonpcouch** [comment]
Thanks for the awesome package, @thisisnic. Just wanted to drop a note that I tested this feature on the vignettes in a few packages in the tidymodels and it worked great—the reporting is really helpfully formatted and I don't see any false positives or false negatives. :-)

**Comment by thisisnic** [comment]
Thanks @simonpcouch
